File: 106_Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal/106_1.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x): val(x), left(nullptr), right(nullptr) { }
};

class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        if (inorder.size() == 0) return nullptr;
        vector<int> left_in, right_in, left_post, right_post;
        TreeNode *root = new TreeNode(postorder.back());
        int mid;
        for (mid = 0; mid < inorder.size(); ++mid)
            if (inorder[mid] == postorder.back())
                break;

        for (int i = 0; i < mid; ++i) {
            left_in.emplace_back(inorder[i]);
            left_post.emplace_back(postorder[i]);
        }

        for (int i = mid + 1; i < inorder.size(); ++i) {
            right_in.emplace_back(inorder[i]);
            right_post.emplace_back(postorder[i - 1]);
        }

        root->left  = buildTree(left_in, left_post);
        root->right = buildTree(right_in, right_post);
        return root;
    }

    static void traverse(TreeNode *root) {
        if (!root) return;
        cout << root->val << " ";
        traverse(root->left);
        traverse(root->right);
    }
};

int main() {
    vector<int> inorder = {9, 3, 15, 20, 7};
    vector<int> postorder = {9, 15, 7, 20, 3};
    Solution s;
    Solution::traverse(s.buildTree(inorder, postorder));
}

File: 106_Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal/106_1.py
from typing import List, Dict


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        if not inorder:
            return
        
        def buildBranch(last, left, right):
            if left > right:
                return
            mid = inorder.index(postorder[last])
            root = TreeNode(postorder[last])
            root.left  = buildBranch(last - (right - mid) - 1, left, mid - 1)
            root.right = buildBranch(last - 1, mid + 1, right)
            return root
        return buildBranch(len(postorder) - 1, 0, len(inorder) - 1)

def traverse(tree: TreeNode) -> None:
    if not tree:
        return
    print(tree.val, " ", end=" ")
    traverse(tree.left)
    traverse(tree.right)
        

if __name__ == "__main__":
    s = Solution()
    preorder = [3, 9, 1, 2, 20, 15, 7]
    inorder = [1, 9, 2, 3, 15, 20, 7]
    traverse(s.buildTree(preorder, inorder))
